runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -216,7 +245,123 @@ function M.on_publish_diagnostics(_, result, ctx, config)
   end
 
   client_id = get_client_id(client_id)
-  local namespace = M.get_namespace(client_id)
+  local namespace = M.get_namespace(client_id, false)
+
+  if config then
+    for _, opt in pairs(config) do
+      if type(opt) == 'table' then
+        if not opt.severity and opt.severity_limit then
+          opt.severity = { min = severity_lsp_to_vim(opt.severity_limit) }
+        end
+      end
+    end
+
+    -- Persist configuration to ensure buffer reloads use the same
+    -- configuration. To make lsp.with configuration work (See :help
+    -- lsp-handler-configuration)
+    vim.diagnostic.config(config, namespace)
+  end
+
+  vim.diagnostic.set(namespace, bufnr, diagnostic_lsp_to_vim(diagnostics, bufnr, client_id))
+end
+
+---@private
+local function resolve_bufnr(bufnr)
+  return bufnr > 0 and bufnr or api.nvim_get_current_buf()
+end
+
+--- Refresh diagnostics for a buffer
+---
+---@param opts (nil|table) Optional arguments
+---  - bufnr (integer, default: 0): Buffer whose hints to refresh
+---  - only_visible (boolean, default: false): Whether to only refresh diagnostics for the visible regions of the buffer
+---
+---@private
+local function refresh(opts)
+  opts = opts or {}
+  local bufnr = resolve_bufnr(opts.bufnr or 0)
+  local only_visible = opts.only_visible or false
+  local buffer_windows = {}
+  for _, winid in ipairs(api.nvim_list_wins()) do
+    if api.nvim_win_get_buf(winid) == bufnr then
+      table.insert(buffer_windows, winid)
+    end
+  end
+  for _, window in ipairs(buffer_windows) do
+    local first = vim.fn.line('w0', window)
+    local last = vim.fn.line('w$', window)
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = first - 1, character = 0 },
+        ['end'] = { line = last, character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/diagnostic', params)
+  end
```

Comment:
```suggestion
  for _, winid in ipairs(api.nvim_list_wins()) do
    if api.nvim_win_get_buf(winid) == bufnr then
      local first = vim.fn.line('w0', window)
      local last = vim.fn.line('w$', window)
      local params = {
        textDocument = util.make_text_document_params(bufnr),
        range = {
          start = { line = first - 1, character = 0 },
          ['end'] = { line = last, character = 0 },
        },
      }
      vim.lsp.buf_request(bufnr, 'textDocument/diagnostic', params)
    end
  end
```

Looks like the intermediate `buffer_windows` table is not needed?

Reply:
good point. we should clean up `inlay_hint.lua` as well, that's where I stole this code from.
or refactor this kind of function out, since they're doing the same thing except for the LSP request method.